File: kotlin-node/karakum/src/jsMain/resources/plugins/PromiseFunctionApiPlugin.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

function isPromiseType(node) {
    return (
        ts.isTypeReferenceNode(node)
        && ts.isIdentifier(node.typeName)
        && node.typeName.text.endsWith("Promise")
    )
}

function isPromiseFunction(node) {
    return (
        ts.isFunctionDeclaration(node)

        && node.type
        && isPromiseType(node.type)
    )
}

function isConflictingOverload(node, signature) {
    return (
        (
            node.name.text === "mkdir"

            && (
                signature.length === 1
                || (
                    signature[1]
                    && ts.isTypeReferenceNode(signature[1].type)
                    && ts.isIdentifier(signature[1].type.typeName)
                    && signature[1].type.typeName.text === "Mode"
                )
            )

            && node.type
            && isPromiseType(node.type)
            && ts.isUnionTypeNode(node.type.typeArguments[0])
        )
        || (
            (
                node.name.text === "mkdtemp"
                || node.name.text === "realpath"
            )

            && (
                signature.length === 1
                || (
                    signature[1]
                    && ts.isTypeReferenceNode(signature[1].type)
                    && ts.isIdentifier(signature[1].type.typeName)
                    && (
                        signature[1].type.typeName.text === "ObjectEncodingOptions"
                        || signature[1].type.typeName.text === "BufferEncoding"
                    )
                )
            )

            && node.type
            && isPromiseType(node.type)
            && ts.isUnionTypeNode(node.type.typeArguments[0])
        )
        || (
            node.name.text === "readlink"

            && (
                signature.length === 1
                || (
                    signature[1]
                    && ts.isTypeReferenceNode(signature[1].type)
                    && ts.isIdentifier(signature[1].type.typeName)
                    && signature[1].type.typeName.text === "ObjectEncodingOptions"
                )
            )

            && node.type
            && isPromiseType(node.type)
            && ts.isUnionTypeNode(node.type.typeArguments[0])
        )
        || (
            node.name.text === "readdir"

            && (
                signature.length === 1
                || (
                    signature[1]
                    && (
                        (
                            ts.isTypeReferenceNode(signature[1].type)
                            && ts.isIdentifier(signature[1].type.typeName)
                            && signature[1].type.typeName.text === "BufferEncoding"
                        )
                        || (
                            ts.isIntersectionTypeNode(signature[1].type)
                            && signature[1].type.types[0]
                            && ts.isTypeReferenceNode(signature[1].type.types[0])
                            && ts.isIdentifier(signature[1].type.types[0].typeName)
                            && signature[1].type.types[0].typeName.text === "ObjectEncodingOptions"
                        )
                    )
                )
            )

            && node.type
            && isPromiseType(node.type)
            && ts.isUnionTypeNode(node.type.typeArguments[0])
        )
    )
}

export default {
    setup(context) {
        this.promiseApiDeclarations = []
    },

    traverse() {
    },

    render(node, context, next) {
        const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

        if (
            isPromiseFunction(node)
            && node.name
        ) {
            const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

            const namespace = typeScriptService?.findClosest(node, ts.isModuleDeclaration)

            const name = node.name.text

            const typeParameters = node.typeParameters
                ?.map(typeParameter => next(typeParameter))
                ?.join(", ")

            const returnType = next(node.type)
            let returnTypePayload = next(node.type.typeArguments[0])
            if (
                ts.isTypeReferenceNode(node.type)
                && ts.isIdentifier(node.type.typeName)
                && node.type.typeName.text === "PipelinePromise"
            ) {
                returnTypePayload = "Any?"
            }

            const body = karakum.convertParameterDeclarations(node, context, next, {
                strategy: "function",
                template: (parameters, signature) => {
                    if (isConflictingOverload(node, signature)) {
                        return ""
                    }

                    return `
@seskar.js.JsAsync
external suspend fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}(${parameters})${karakum.ifPresent(returnTypePayload, it => `: ${it}`)}
                    `
                }
            })

            const nodeInfo = {
                sourceFileName,
                namespace,
                fileName: `${name}.suspend.kt`,
                body,
            }

            this.promiseApiDeclarations.push(nodeInfo)

            return karakum.convertParameterDeclarations(node, context, next, {
                strategy: "function",
                template: (parameters, signature) => {
                    if (isConflictingOverload(node, signature)) {
                        return ""
                    }

                    return `
@JsName("${node.name.text}")
external fun ${karakum.ifPresent(typeParameters, it => `<${it}> `)}${name}Async(${parameters})${karakum.ifPresent(returnType, it => `: ${it}`)}
                    `
                }
            })
        }

        return null
    },

    generate(context) {
        return karakum.generateDerivedDeclarations(this.promiseApiDeclarations, context);
    },
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/PromiseFunctionApiPlugin.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.*
import io.github.sgrishchenko.karakum.extension.plugins.*
import io.github.sgrishchenko.karakum.extension.plugins.Signature
import io.github.sgrishchenko.karakum.structure.derived.DerivedDeclaration
import io.github.sgrishchenko.karakum.structure.derived.generateDerivedDeclarations
import io.github.sgrishchenko.karakum.util.escapeIdentifier
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import js.array.ReadonlyArray
import typescript.*

private fun isPromiseType(node: Node) = nullable {
    ensure(isTypeReferenceNode(node))

    val typeName = node.typeName
    ensure(isIdentifier(typeName))
    ensure(typeName.text.endsWith("Promise"))
} != null

private fun isPromiseFunction(node: Node) = nullable {
    ensure(isFunctionDeclaration(node))

    val type = ensureNotNull(node.type)
    ensure(isPromiseType(type))
} != null

private fun isConflictingOverload(node: FunctionDeclaration, signature: Signature) = nullable {
    val name = ensureNotNull(node.name)

    nullable {
        ensure(name.text == "mkdir")

        val type = ensureNotNull(node.type)
        ensure(isTypeReferenceNode(type))
        ensure(isPromiseType(type))

        val typeArguments = ensureNotNull(type.typeArguments)
        ensure(isUnionTypeNode( typeArguments.asArray().first()))

        nullable {
            ensure(signature.size == 1)
        } ?: nullable {
            val secondParameter = ensureNotNull(signature.getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isTypeReferenceNode(secondParameterType))

            val secondParameterTypeName = secondParameterType.typeName
            ensure(isIdentifier(secondParameterTypeName))
            ensure(secondParameterTypeName.text == "Mode")
        }
    } ?: nullable {
        ensure(
            name.text == "mkdtemp"
                    || name.text == "realpath"
        )

        val type = ensureNotNull(node.type)
        ensure(isTypeReferenceNode(type))
        ensure(isPromiseType(type))

        val typeArguments = ensureNotNull(type.typeArguments)
        ensure(isUnionTypeNode( typeArguments.asArray().first()))

        nullable {
            ensure(signature.size == 1)
        } ?: nullable {
            val secondParameter = ensureNotNull(signature.getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isTypeReferenceNode(secondParameterType))

            val secondParameterTypeName = secondParameterType.typeName
            ensure(isIdentifier(secondParameterTypeName))
            ensure(
                secondParameterTypeName.text == "ObjectEncodingOptions"
                        || secondParameterTypeName.text == "BufferEncoding"
            )
        }
    } ?: nullable {
        ensure(name.text == "readlink")

        val type = ensureNotNull(node.type)
        ensure(isTypeReferenceNode(type))
        ensure(isPromiseType(type))

        val typeArguments = ensureNotNull(type.typeArguments)
        ensure(isUnionTypeNode( typeArguments.asArray().first()))

        nullable {
            ensure(signature.size == 1)
        } ?: nullable {
            val secondParameter = ensureNotNull(signature.getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isTypeReferenceNode(secondParameterType))

            val secondParameterTypeName = secondParameterType.typeName
            ensure(isIdentifier(secondParameterTypeName))
            ensure(secondParameterTypeName.text == "ObjectEncodingOptions")
        }
    } ?: nullable {
        ensure(name.text == "readdir")

        val type = ensureNotNull(node.type)
        ensure(isTypeReferenceNode(type))
        ensure(isPromiseType(type))

        val typeArguments = ensureNotNull(type.typeArguments)
        ensure(isUnionTypeNode( typeArguments.asArray().first()))

        nullable {
            ensure(signature.size == 1)
        } ?: nullable {
            val secondParameter = ensureNotNull(signature.getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isTypeReferenceNode(secondParameterType))

            val secondParameterTypeName = secondParameterType.typeName
            ensure(isIdentifier(secondParameterTypeName))
            ensure(secondParameterTypeName.text == "BufferEncoding")
        } ?: nullable {
            val secondParameter = ensureNotNull(signature.getOrNull(1))

            val secondParameterType = ensureNotNull(secondParameter.type)
            ensure(isIntersectionTypeNode(secondParameterType))

            val secondParameterTypeFirstType = secondParameterType.types.asArray().first()
            ensure(isTypeReferenceNode(secondParameterTypeFirstType))

            val secondParameterTypeFirstTypeName = secondParameterTypeFirstType.typeName
            ensure(isIdentifier(secondParameterTypeFirstTypeName))
            ensure(secondParameterTypeFirstTypeName.text == "ObjectEncodingOptions")
        }
    }
} != null

class PromiseFunctionApiPlugin : Plugin {
    private val promiseApiDeclarations = mutableListOf<DerivedDeclaration>()

    override fun setup(context: Context) = Unit

    override fun traverse(node: Node, context: Context) = Unit

    override fun render(node: Node, context: Context, next: Render<Node>) = nullable {
        val sourceFileName = node.getSourceFileOrNull()?.fileName ?: "generated.d.ts"

        val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

        val namespace = typeScriptService.findClosestNamespace(node)

        ensure(isFunctionDeclaration(node))
        ensure(isPromiseFunction(node))

        val nameNode = ensureNotNull(node.name)

        val name = escapeIdentifier(next(nameNode))

        val typeParameters = node.typeParameters?.asArray()
            ?.joinToString(", ") { next(it) }

        val returnType = node.type ?.let { next(it) }

        val type = requireNotNull(node.type)
        require(isTypeReferenceNode(type))

        val typeArguments = requireNotNull(type.typeArguments)

        var returnTypePayload = next(typeArguments.asArray().first())

        val typeName = type.typeName

        if (isIdentifier(typeName) && typeName.text == "PipelinePromise") {
            returnTypePayload = "Any?"
        }

        val body = convertParameterDeclarations(
            node, context, next,
            ParameterDeclarationsConfiguration(
                strategy = ParameterDeclarationStrategy.function,
                template = template@{ parameters, signature ->
                    if (isConflictingOverload(node, signature)) return@template ""

                    """
                        @seskar.js.JsAsync
                        external suspend fun ${ifPresent(typeParameters) { "<${it}> " }}${name}(${parameters})${ifPresent(returnTypePayload) { ": $it"}}
                    """.trimIndent()
                }
            )
        )

        val nodeInfo = DerivedDeclaration(
            sourceFileName,
            namespace,
            fileName = "${name}.suspend.kt",
            body,
        )

        promiseApiDeclarations += nodeInfo

        convertParameterDeclarations(
            node, context, next,
            ParameterDeclarationsConfiguration(
                strategy = ParameterDeclarationStrategy.function,
                template = template@{ parameters, signature ->
                    if (isConflictingOverload(node, signature)) return@template ""

                    """
                        @JsName("$name")
                        external fun ${ifPresent(typeParameters) { "<${it}> " }}${name}Async(${parameters})${ifPresent(returnType) { ": $it" }}
                    """.trimIndent()
                }
            )
        )
    }

    override fun generate(context: Context, render: Render<Node>): ReadonlyArray<GeneratedFile> {
        return generateDerivedDeclarations(promiseApiDeclarations.toTypedArray(), context)
    }
}
